toolchain/check/eval.cpp

```diff
@@ -1549,641 +1544,672 @@ static auto MakeFacetTypeResult(Context& context,
       phase);
 }
 
-// Implementation for `TryEvalInst`, wrapping `Context` with `EvalContext`.
+// The result of constant evaluation of an instruction.
+class ConstantEvalResult {
+ public:
+  // Produce a new constant as the result of an evaluation. The phase of the
+  // produced constant must be the same as the phase of the operands in the
+  // evaluation.
+  static auto New(SemIR::Inst inst) -> ConstantEvalResult {
+    return ConstantEvalResult(inst);
+  }
+
+  // Produce an existing constant as the result of an evaluation.
+  static constexpr auto Existing(SemIR::ConstantId existing_id)
+      -> ConstantEvalResult {
+    CARBON_CHECK(existing_id.is_constant());
+    return ConstantEvalResult(existing_id);
+  }
+
+  // Indicates that an error was produced by evaluation.
+  static const ConstantEvalResult Error;
+
+  // Indicates that we encountered an instruction whose evaluation is
+  // non-constant despite having constant operands. This should be rare;
+  // usually we want to produce an error in this case.
+  static const ConstantEvalResult NotConstant;
+
+  // Indicates that we encountered an instruction for which we've not
+  // implemented constant evaluation yet. Instruction is treated as not
+  // constant.
+  static const ConstantEvalResult TODO;
+
+  // Returns whether the result of evaluation is that we should produce the
+  // instruction itself as a new constant.
+  auto is_new() const -> bool { return !result_id_.has_value(); }
+
+  // Returns whether the result of evaluation is an existing constant.
```

Comment:
"Returns whether..."  but this function doesn't return a `bool`.

Reply:
Done.